### ISSUE TITLE
cblas3 and zblas3 tests are failing on Broadwell and Skylake with GCC 11.1.0

### ISSUE THREAD

**Issue by Chaosit** [xianyi/OpenBLAS#3272]
I have tried building OpenBLAS v. 0.3.15 on my workstation with Intel Xeon E5-2687W v4 and GCC-11.1.0 and some of the tests, namely cblas3 and zblas3 seem to be failing with a completely wrong result. This is reproducible also on my Skylake i7 CPU at home. The error looks like this:

```
OMP_NUM_THREADS=2 ./xzcblat3 < zin3
TESTS OF THE COMPLEX*16 LEVEL 3 BLAS

THE FOLLOWING PARAMETER VALUES WILL BE USED:
FOR N 0 1 2 3 5 9 35
FOR ALPHA ( 0.0, 0.0) ( 1.0, 0.0) ( 0.7,-0.9)
FOR BETA ( 0.0, 0.0) ( 1.0, 0.0) ( 1.3,-1.1)

ROUTINES PASS COMPUTATIONAL TESTS IF TEST RATIO IS LESS THAN 16.00

COLUMN-MAJOR AND ROW-MAJOR DATA LAYOUTS ARE TESTED

RELATIVE MACHINE PRECISION IS TAKEN TO BE 2.2E-16

cblas_zgemm PASSED THE TESTS OF ERROR-EXITS

******* FATAL ERROR - COMPUTED RESULT IS LESS THAN HALF ACCURATE *******
EXPECTED RESULT COMPUTED RESULT
1 ( 0.495928E-01, 0.197144 ) ( -0.176908 , -0.293573E-01)
2 ( -0.154221 , -0.169237 ) ( 0.179134 , 0.164118 )
******* cblas_zgemm FAILED ON CALL NUMBER:
8644: cblas_zgemm ( CblasColMajor,CblasConjTrans, CblasNoTrans,
2, 1, 1, ( 1.0, 0.0) , A, 2, B, 2, ( 0.0, 0.0) , C, 3).
******* cblas_zgemm FAILED ON CALL NUMBER:
4537: cblas_zgemm ( CblasRowMajor, CblasNoTrans, CblasNoTrans,
1, 1, 0, ( 0.0, 0.0) , A, 2, B, 1, ( 0.0, 0.0) , C, 2)
```

**Comment by martin-frbg** [comment]
Both should be the same cpu target (Haswell) as far as OpenBLAS is concerned. I cannot reproduce this with either gcc 10.2 or a recent snapshot of gcc12 right away, will need to install 11.1 to check. Can you please make sure that gcc and gfortran are the same version, and the cpu firmware is reasonably current ?

**Comment by martin-frbg** [comment]
Not reproduced on i7-8700K with gcc-11.1 and either 0.3.15 or current develop branch.

**Comment by martin-frbg** [comment]
Did you specify any non-default build options (I checked both with and without USE_OPENMP=1) ? 
Also, one possible pitfall when using a non-default installation of gcc on a system is that while `gcc` may already point to the "new" compiler, `cc` still invokes the system compiler. One can use the `CC` argument to make/cmake to specify which compiler to use (as in `make CC=gcc` or `make CC=/opt/gcc11/bin/gcc` etc).

**Comment by Chaosit** [comment]
Thanks a lot for the reply! In fact, today I was not able to reproduce this problem on my core i7 either, although it still persists on my workstation with Xeon. I did check that `gcc` and `gfortran` are of the same version:
```
gcc --version
gcc (GCC) 11.1.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

gfortran --version
GNU Fortran (GCC) 11.1.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
```

I have also checked the microcode version:

```
sudo dmesg | grep 'microcode'
[    0.000000] microcode: microcode updated early to revision 0xb000038, date = 2019-06-18
[    2.930589] microcode: sig=0x406f1, pf=0x1, revision=0xb000038
[    2.932025] microcode: Microcode Update Driver: v2.2.
```
Not sure if that is reasonably new, though.

I have also used only default `cmake` options setting `CMAKE_BUILD_TYPE` to `Release`: `cmake -DCMAKE_C_COMPILER=gcc -DCMAKE_Fortran_COMPILER=gfortran -DCMAKE_BUILD_TYPE=Release ..`

**Comment by martin-frbg** [comment]
Somewhat disconcerting. but not reproducible on my systems even after repeated compiles (tried both gmake and cmake builds)

**Comment by cyyever** [comment]
I have the same test errors on gcc11

**Comment by martin-frbg** [comment]
What is your hardware @cyyever ? (OS is some BSD if I remember correctly)

**Comment by cyyever** [comment]
archlinux + gcc11
cpu is (fetchted from Windows) 
Caption                                 DeviceID  MaxClockSpeed  Name                                      NumberOfCores  Status
Intel64 Family 6 Model 158 Stepping 10  CPU0      2592           Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz  6              OK

**Comment by Chaosit** [comment]
Just in case, for me everything works fine as long as I do not use GCC 11, e.g. GCC 10.x works perfectly fine for me.

**Comment by martin-frbg** [comment]
Saw this now with current develop and a gcc11.1 built from source, but only when building with cmake and setting CMAKE_BUILD_TYPE=Release, could be some kind of compiler option mixup in the cmake build files.

**Comment by martin-frbg** [comment]
As build type "Release" selects `-O3` rather than the more conservative `-O2` used in gmake builds, this is possibly "just" an optimizer bug in 11.1, or in any case something that could be worked around with a `#pragma GCC optimize "O2"` in the affected file(s)

**Comment by martin-frbg** [comment]
Interestingly it appears to be the ctest3 and ztest3 programs themselves that fail when compiled at `-O3` by gcc 11.1 or a development snapshot of 12.

**Comment by martin-frbg** [comment]
Problem goes away with adding `!GCC$ novector` directives to the loops that calculate the reference values in both tests. (Unfortunately this directive is not known to older versions of gfortran and treated as an error by at least  v7.5.0)

**Comment by martin-frbg** [comment]
`-fno-tree-vectorize` achieves the same and is understood by older versions. Still need to trim this down to something that can serve as a reproducer for a gcc bug report

**Comment by martin-frbg** [comment]
Possibly GCC bug 10125 which just got fixed with GCC 11.2 (but that was in gcc and the workaround of `-fno-tree-loop-im` does not work here, so will need to build 11.2 and retest with that)
Update: bug still present in 11.2, but while preparing to bisect gcc I found that I cannot reproduce the problem with a checkout from yesterday (and I keep hitting "some good revs are not ancestors of the bad rev" in my attempts to bisect, so may abandon this soon)
